pkg/cosign/verify.go

```diff
@@ -1281,3 +1292,57 @@ func correctAnnotations(wanted, have map[string]interface{}) bool {
 	}
 	return true
 }
+
+// verifyImageSignaturesExperimentalOCI does all the main cosign checks in a loop, returning the verified signatures.
+// If there were no valid signatures, we return an error, using OCI 1.1+ behavior.
+func verifyImageSignaturesExperimentalOCI(ctx context.Context, signedImgRef name.Reference, co *CheckOpts) (checkedSignatures []oci.Signature, bundleVerified bool, err error) {
+	// Enforce this up front.
+	if co.RootCerts == nil && co.SigVerifier == nil {
+		return nil, false, errors.New("one of verifier or root certs is required")
+	}
+
+	// This is a carefully optimized sequence for fetching the signatures of the
+	// entity that minimizes registry requests when supplied with a digest input
+	digest, err := ociremote.ResolveDigest(signedImgRef, co.RegistryClientOpts...)
+	if err != nil {
+		return nil, false, err
+	}
+	h, err := v1.NewHash(digest.Identifier())
+	if err != nil {
+		return nil, false, err
+	}
+
+	var sigs oci.Signatures
+	sigRef := co.SignatureRef
+	if sigRef == "" {
+		artifactType := ociexperimental.ArtifactType("sig")
+		index, err := ociremote.Referrers(digest, artifactType, co.RegistryClientOpts...)
+		if err != nil {
+			return nil, false, err
+		}
+		results := index.Manifests
+		numResults := len(results)
+		if numResults == 0 {
+			return nil, false, fmt.Errorf("unable to locate reference with artifactType %s", artifactType)
+		} else if numResults > 1 {
+			// TODO: if there is more than 1 result.. what does that even mean?
+			fmt.Printf("WARNING: there were a total of %d references with artifactType %s\n", numResults, artifactType)
```

Comment:
Maybe: "expected only 1" in the warning?

Reply:
This isn't a terribly actionable warning right now sadly. I guess because we don't know why the heck it would happen, ha